### PR TITLE
Attempt at pseudo-fixing the performance problems

### PR DESCRIPTION
**Pull request by Randshot**
#422 #433 
This is my attempt at fixing some of the performance issues when the desired object type isn't available.
For more detailed information, look into the changed files.

Please tell me your opinions about this. ( I know that it isn't very elegant ;) )

### CODE REVIEW

File: Assets/Scripts/Models/Character.cs [hunk 1]
```diff
@@ -322,7 +336,7 @@ bool CheckForJobMaterials()
                     {
                         //Logger.Log("pathAStar is null and we have no path to object of type: " + desired.objectType);
                         // Cancel the job, since we have no way to get any raw materials!
-                        Logger.Log("No tile contains objects of type '" + desired.objectType + "' to satisfy job requirements.");
+                        //Logger.Log("No tile contains objects of type '" + desired.objectType + "' to satisfy job requirements.");
```
Comment (zwrawr) [inline Assets/Scripts/Models/Character.cs]: Instead of commenting that out you could do Logger.VerboseLog()

**Comment by Mizipzor** [comment]
Spaces, not tabs. :)

**Comment by Randshot** [comment]
@zwrawr 
I will try that :)

@Mizipzor 
Oops. I used the builtin editor. It seems like that one uses tabs instead of spaces.
I will correct it shortly.

**Comment by petroVARS** [comment]
I am working on the same problem, but without using a timer. Maybe wait with accepting this for a bit.

**Comment by Randshot** [comment]
@zwrawr Unity says, that there is no Logger.VerboseLog() ...

**Comment by Randshot** [comment]
So, here are a couple comparisons of the old and new versions.
(Sry for the large images, don't know how to scale them down.)

**Old Version**
_1 Chars; 1 Jobs_
orig1jobs1chars
_1 Chars; 5 Jobs_
orig5jobs2chars
_4 Chars; 4 Jobs_
orig4jobs4chars
_4 Chars; 5 Jobs_
orig5jobs4chars
-----------------

**New Version**
_1 Chars; 1 Jobs_
new1jobs1chars
_1 Chars; 5 Jobs_
new5jobs2chars
_4 Chars; 4 Jobs_
new4jobs4chars
_4 Chars; 5 Jobs_
new5jobs4chars

File: Assets/Scripts/Models/Character.cs [hunk 2]
```diff
@@ -121,6 +121,9 @@ public Tile JobTile
     /// Our job, if any.
     Job myJob;
 
+	/// Job wait timer (to lessen material checks)
+	int jobWaitTimer = 0;
```
Comment (Dormanil) [inline Assets/Scripts/Models/Character.cs]: I think you missed some tabs here. ;)

File: Assets/Scripts/Models/Character.cs [hunk 3]
```diff
@@ -191,29 +194,31 @@ void GetNewJob()
...
-        {
-            GetNewJob();
-        }
-
-        if (CheckForJobMaterials()) //make sure all materials are in place
-        {
-            // If we get here, then the job has all the material that it needs.
-            // Lets make sure that our destination tile is the job site tile.
-            DestTile = JobTile;
+        if (jobWaitTimer > 0) {
```
Comment (Dormanil) [inline Assets/Scripts/Models/Character.cs]: Braces should always be on new lines. Always.

File: Assets/Scripts/Models/Character.cs [hunk 4]
```diff
@@ -191,29 +194,31 @@ void GetNewJob()
...
 
-            // Are we there yet?
-            if (CurrTile == DestTile)
-            {
-                // We are at the correct tile for our job, so 
-                // execute the job's "DoWork", which is mostly
-                // going to countdown jobTime and potentially
-                // call its "Job Complete" callback.
-                myJob.DoWork(deltaTime);
+            if (CheckForJobMaterials ()) { //make sure all materials are in place
```
Comment (Dormanil) [inline Assets/Scripts/Models/Character.cs]: Comments are on new lines as well, have a space between `//` and the text, the text starts with a capital letter, and ends with a period.

**Comment by Randshot** [comment]
Anything else ;)

**Comment by petroVARS** [comment]
I implemented a pause function for jobs. Every job that does not have the required materials will not be inserted into the queue, but into a waiting list instead. Whenever a new inventory is created, the list is checked whether a job can be restarted (there was already a callback for this). I think this should at least be a more elegant version instead of using a timer. 

Regarding performance, I am not sure if there is any notable difference to your version. For me, the fps counter seems to be bugged, I often get only 30 fps while absolutely nothing is going on while the stats window shows around 300 fps. Anyway, the fps does not drop significantly anymore. I logged every A* path call and it appears that only the minimum number of calls are made. I will make a pull request after i got some sleep and tested it a bit more .

**Comment by Randshot** [comment]
@petroVARS I find this very amusing since I also thought of doing something similar to this xD
I didn't know about the already existing callback function though.

**Comment by Randshot** [comment]
@petroVARS I think I know why you only get 30 fps. You may have the VSync option 'Every Blank' enabled?

**Comment by vogonistic** [comment]
@Randshot I'm accepting this PR now, but I want you to make a follow up PR that replaces `jobWaitTimer` with the creation of a Waiting job instead. Also replace `JobQueue.QuickCheck` with something that check's if the key exists but is of length 0, since if that material has ever existed, `QuickCheck` will always say true.

**Comment by Mizipzor** [comment]
@vogonistic Im still working my way through the notifications, but if there isnt already a pull request for that could we create an issue just so it isnt forgotten?

**Comment by vogonistic** [comment]
I only created #482 for now.

**Pull request merged**